src/ConnectedMode/Binding/ImportBeforeFileGenerator.cs

```diff
@@ -0,0 +1,105 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.IO.Abstractions;
+using System.Text;
+using System.Xml;
+
+namespace SonarLint.VisualStudio.ConnectedMode.Binding
+{
+    internal class ImportBeforeFileGenerator
+    {
+        private readonly IFileSystem fileSystem;
+
+        private const string targetsFileName = "SonarLint.targets";
+
+        public ImportBeforeFileGenerator() : this(new FileSystem()) { }
+
+        public ImportBeforeFileGenerator(IFileSystem fileSystem)
+        {
+            this.fileSystem = fileSystem;
+
+            CreateTargetsFileIfNotExists();
+        }
+
+        private void CreateTargetsFileIfNotExists()
+        {
+            var sb = new StringBuilder();
+            var xmlWriter = Create(sb);
+
+            xmlWriter.WriteStartElement("Project", "http://schemas.microsoft.com/developer/msbuild/2003");
+            xmlWriter.WriteAttributeString("xmlns", "http://schemas.microsoft.com/developer/msbuild/2003");
+
+            xmlWriter.WriteStartElement("ItemGroup");
+
+            xmlWriter.WriteEndElement();
+            xmlWriter.WriteEndElement();
+
+            xmlWriter.Close();
+            WriteFile(sb.ToString());
+        }
+
+        private void WriteFile(string xml)
+        {
+            var pathToImportBefore = GetPathToImportBefore();
+
+            if (!fileSystem.Directory.Exists(pathToImportBefore))
+            {
+                fileSystem.Directory.CreateDirectory(pathToImportBefore);
+            }
+
+            var fullPath = Path.Combine(pathToImportBefore, targetsFileName);
+
+            if (fileSystem.File.Exists(fullPath) && fileSystem.File.ReadAllText(fullPath) == xml)
```

Comment:
I think the current approach is fine. Only writing the file when we need to (i.e. missing or different) reduces the chances of us trying to write the file at the same time that is being read on another thread.

Having said that, pretty much any file operation could fail, so it would be worth wrapping all of the IO calls in an exception handler to catch and log any errors. And if we are having to add logging, we might as well add some verbose logging to help diagnose the other code paths i.e. directory being created, file already exists etc. (I wouldn't add tests for the verbose logging though)